### PR TITLE
Feature: Pattern Metadata

### PR DESCRIPTION
**Pull request by lyzadanger**
This PR represents a complete rewrite of pattern and collection parsing to make pattern-file-specific metadata for ordering and hiding patterns within collections work. 

It is now possible to manage collection metadata, as it was before, using a `collection.[yml|json]` file in the pattern directory, a la:

```
name: My Special Patterns Collection
order:
  - fruit
  - flies
  - like-a
  - banana
hidden:
  - dignity
```

but it is also possible to adjust this on a pattern-file level, e.g. `banana.hbs`:

```
---
name: You Think I am A Banana
order: 5
hidden: true
---
```

It's smart enough to figure out gaps and missing patterns. Pattern-level data will supersede collection data. There are tests, but I'd like to write some more when I have more time, and it could use some more better error-handling.

Enjoy.

/cc @mrgerardorodriguez @erikjung @saralohr @nicolemors

### CODE REVIEW

**Comment by tylersticka** [comment]
This makes me a happy camper. Thank you! :smiley:

**Comment by lyzadanger** [comment]
Could someone be so kind as to review this PR for me? /cc @tylersticka @mrgerardorodriguez @erikjung

File: src/parse/patterns.js [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+ * @module parse/patterns
+ */
+import { resourceId, resourceKey } from '../utils/object';
+import { readFiles, readFileTree } from '../utils/parse';
+import { titleCase } from '../utils/shared';
+import path from 'path';
+
+const isPattern = obj => obj.hasOwnProperty('path');
+const collectionPath = itms => path.dirname(itms[Object.keys(itms).pop()].path);
+const collectionKey = itms => collectionPath(itms).split(path.sep).pop();
```
Comment (erikjung) [inline src/parse/patterns.js]: I like these functions :+1:

**Comment by erikjung** [comment]
:+1:

**Comment by mrgerardorodriguez** [comment]
LGTM! :+1:

**Pull request merged**